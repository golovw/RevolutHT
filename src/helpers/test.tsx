import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from 'styles';

export const wrapWithTheme = (Component: React.ReactNode) => (
  <ThemeProvider theme={theme}>
    {Component}
  </ThemeProvider>
);

export const renderWithTheme = (Component: React.ReactNode, options?: any) => render(wrapWithTheme(Component), options);
