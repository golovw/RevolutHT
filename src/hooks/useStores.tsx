import React from 'react';

import stores from 'stores';

export const useStores = () => React.useContext(stores);
