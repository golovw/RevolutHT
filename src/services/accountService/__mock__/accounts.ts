export const accountsMock = [
  {
    id: '33bba04a',
    currency: 'USD',
    balance: '2340.00',
  },
  {
    id: '33bba31a',
    currency: 'EUR',
    balance: '87.00'
  },
  {
    id: '33bba428',
    currency: 'GBP',
    balance: '104.00'
  },
];
