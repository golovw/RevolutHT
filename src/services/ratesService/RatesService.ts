import config from 'config';

import { ApiAdapterI } from 'adapters/apiAdapter';

const {
  api: {
    rates: {
      domain,
      path,
      appId,
      appIdParam,
    },
  },
} = config;

export class RatesService {
  apiAdapter;

  constructor(apiAdapter: ApiAdapterI) {
    this.apiAdapter = apiAdapter;
  }

  getRates = async () => {
    try {
      const { data } = await this.apiAdapter.get(`${domain}${path}`, { [appIdParam]: appId });
      return data;
    } catch(e) {
      console.error(e);
    }
  }

  getRatesRecursively = async () => {
    try {
      const { data } = await this.apiAdapter.get(`${domain}${path}`, { [appIdParam]: appId });
      return data;
    } catch(e) {
      console.error(e);
    }
  }
}
