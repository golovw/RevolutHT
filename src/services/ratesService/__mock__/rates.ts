export const ratesMock = {
  EUR: 0.818746,
  GBP: 0.706283,
  USD: 1,
};
