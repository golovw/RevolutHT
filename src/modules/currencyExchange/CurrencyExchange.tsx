import React, { useCallback, useEffect, useState } from 'react';
import { observer } from 'mobx-react';

import { useStores} from 'hooks/useStores';

import { Account } from './components/account';

import { convert } from 'utils/currency';
import { getRateLabel } from './CurrencyExchange.utils';

import { Account as AccountI } from 'ducks/Account.duck';

import { emptyAccount } from "./CurrencyExchange.constants";

import {
  Container,
  AccountScope,
  AccountScopeGrayed,
  Rate,
  TransactionButton,
} from './CurrencyExchange.styles';

export const CurrencyExchange = observer(() => {
  const {
    accountsStore: {
      accounts,
      getAccounts,
      makeExchange
    },
    ratesStore: {
      rates,
      setGetRatesIntervalRequest,
      clearTimers
    }
  } = useStores();

  useEffect(() => {
    getAccounts();
    setGetRatesIntervalRequest();

    return () => clearTimers();
  }, []);

  const [fromAmount, setFromAmount] = useState('0');
  const [toAmount, setToAmount] = useState('0');
  const [fromAccount, setFromAccount] = useState<AccountI>(emptyAccount);
  const [toAccount, setToAccount] = useState<AccountI>(emptyAccount);

  const resolveCurrencyDuplicates = (
    from: AccountI,
    to: AccountI,
    accounts: AccountI[],
    set: (account: AccountI) => void
  ) => {
    if (from && to && from.id && from.id === to.id) {
      const newAccount = accounts.find(({ id }: AccountI) => id !== from.id);
      newAccount && set(newAccount);
    }
  };

  useEffect(() => {
    const convertedValue = convert(rates[fromAccount.currency], rates[toAccount.currency], +fromAmount);
    const formattedValue = convertedValue === 0 ? convertedValue : convertedValue.toFixed(2)
    setToAmount(`${formattedValue}`);
  }, [fromAmount, fromAccount, toAccount, setToAmount]);

  useEffect(() => {
    if (accounts.length) {
      setFromAccount(accounts[0]);
      setToAccount(accounts[1]);
    }
  }, [accounts, setFromAccount, setToAccount]);

  useEffect(() => {
    resolveCurrencyDuplicates(fromAccount, toAccount, accounts, setToAccount);
  }, [fromAccount]);

  useEffect(() => {
    resolveCurrencyDuplicates(fromAccount, toAccount, accounts, setFromAccount);
  }, [toAccount]);

  const getAccountChangeHandler = (handler: (account: AccountI) => void) => {
    return ({ target: { value } }: React.ChangeEvent<{ value: unknown }>) => {
      const newAccount = accounts.find(({ id }: AccountI) => id === value);
      newAccount && handler(newAccount);
    };
  };

  const exchange = useCallback(() => {
    makeExchange(fromAccount.id, toAccount.id, +fromAmount, +toAmount);
    setFromAmount('0');
    setToAmount('0');
  }, [fromAccount, toAccount, fromAmount, toAmount, makeExchange, setFromAmount, setToAmount]);

  return (
    <Container>
      <AccountScope>
        <Account
          amount={fromAmount}
          account={fromAccount}
          accounts={accounts}
          onAccountChange={getAccountChangeHandler(setFromAccount)}
          onAmountChange={setFromAmount}
        />
      </AccountScope>

      <Rate
        label={
          getRateLabel(
            fromAccount.currency,
            toAccount.currency,
            convert(rates[ fromAccount.currency],
              rates[toAccount.currency],
              1
            )
          )
        }
        variant="outlined"
      />

      <AccountScopeGrayed>
        <Account
          isReceiving
          amount={toAmount}
          account={toAccount}
          accounts={accounts}
          onAccountChange={getAccountChangeHandler(setToAccount)}
          onAmountChange={setToAmount}
        />

        <TransactionButton
          variant="contained"
          color="primary"
          onClick={exchange}
        >
          Exchange
        </TransactionButton>
      </AccountScopeGrayed>
    </Container>
  );
});
