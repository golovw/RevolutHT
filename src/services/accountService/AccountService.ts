import { accountsMock } from './__mock__';

import { ApiAdapterI } from 'adapters/apiAdapter';

export class AccountService {
  apiAdapter;

  constructor(apiAdapter: ApiAdapterI) {
    this.apiAdapter = apiAdapter;
  }

  getAccounts = async () => {
    return { data: accountsMock };
  }
}
