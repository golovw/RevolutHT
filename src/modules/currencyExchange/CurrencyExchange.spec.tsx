import React from 'react';
import { cleanup } from "@testing-library/react";

import { renderWithTheme } from 'helpers/test';

import { CurrencyExchange } from './CurrencyExchange';

import { useStores } from 'hooks/useStores';

import { accountsMock } from 'services/accountService/__mock__';
import { ratesMock } from 'services/ratesService/__mock__';

jest.mock('hooks/useStores', () => {
  return {
    useStores: jest.fn(),
  };
});

describe('CurrencyExchange', () => {

  beforeEach(() => {
    jest.useFakeTimers();

    const accountsStore = {
      accounts: accountsMock,
      getAccounts: jest.fn(),
      makeExchange: jest.fn(),
    };

    const ratesStore = {
      rates: ratesMock,
      getRates: jest.fn(),
      setGetRatesIntervalRequest: jest.fn(),
      clearTimers: jest.fn(),
    };

    (useStores as jest.Mock).mockReturnValue({ accountsStore, ratesStore });
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    cleanup();
  });

  describe('Snapshots', () => {
    it('sending account should match snapshot', () => {
      const component = renderWithTheme(
        <CurrencyExchange />
      );

      expect(component).toMatchSnapshot();
    });
  });
});
