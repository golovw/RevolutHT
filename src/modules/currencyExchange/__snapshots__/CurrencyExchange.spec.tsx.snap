// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyExchange Snapshots sending account should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jSFjdj fJOzHk"
      >
        <div
          class="sc-iCoGMd dfSdNJ"
        >
          <div
            class="sc-gtsrHT mjMEe"
          >
            <div
              class="sc-dlnjwi ffkZxH"
            >
              <div
                class="MuiInputBase-root MuiInput-root sc-bdnxRM eMbhKg"
              >
                <div
                  aria-haspopup="listbox"
                  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                  role="button"
                  tabindex="0"
                >
                  USD
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput"
                  tabindex="-1"
                  value="33bba04a"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <div
                class="MuiInputBase-root MuiInput-root sc-eCApnc jQvmOi"
              >
                <input
                  class="MuiInputBase-input MuiInput-input"
                  type="text"
                  value="0"
                />
              </div>
            </div>
            <div
              class="sc-dlnjwi ffkZxH"
            >
              <span
                class="sc-hKFxyN dimflz"
              >
                Balance: 
                $
                2340.00
              </span>
            </div>
          </div>
        </div>
        <div
          class="MuiChip-root sc-pNWdM ddNSfw MuiChip-outlined"
        >
          <span
            class="MuiChip-label"
          >
            $1 = €0.82
          </span>
        </div>
        <div
          class="sc-iCoGMd sc-fujyAs dfSdNJ gHZHtL"
        >
          <div
            class="sc-gtsrHT cMYXGm"
          >
            <div
              class="sc-dlnjwi ffkZxH"
            >
              <div
                class="MuiInputBase-root MuiInput-root sc-bdnxRM eMbhKg"
              >
                <div
                  aria-haspopup="listbox"
                  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                  role="button"
                  tabindex="0"
                >
                  EUR
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput"
                  tabindex="-1"
                  value="33bba31a"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <div
                class="MuiInputBase-root MuiInput-root sc-eCApnc jQvmOi Mui-disabled Mui-disabled"
              >
                <input
                  class="MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  type="text"
                  value="0"
                />
              </div>
            </div>
            <div
              class="sc-dlnjwi ffkZxH"
            >
              <span
                class="sc-hKFxyN dimflz"
              >
                Balance: 
                €
                87.00
              </span>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-jrsJWt bsREHq MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Exchange
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSFjdj fJOzHk"
    >
      <div
        class="sc-iCoGMd dfSdNJ"
      >
        <div
          class="sc-gtsrHT mjMEe"
        >
          <div
            class="sc-dlnjwi ffkZxH"
          >
            <div
              class="MuiInputBase-root MuiInput-root sc-bdnxRM eMbhKg"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                USD
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="33bba04a"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInput-root sc-eCApnc jQvmOi"
            >
              <input
                class="MuiInputBase-input MuiInput-input"
                type="text"
                value="0"
              />
            </div>
          </div>
          <div
            class="sc-dlnjwi ffkZxH"
          >
            <span
              class="sc-hKFxyN dimflz"
            >
              Balance: 
              $
              2340.00
            </span>
          </div>
        </div>
      </div>
      <div
        class="MuiChip-root sc-pNWdM ddNSfw MuiChip-outlined"
      >
        <span
          class="MuiChip-label"
        >
          $1 = €0.82
        </span>
      </div>
      <div
        class="sc-iCoGMd sc-fujyAs dfSdNJ gHZHtL"
      >
        <div
          class="sc-gtsrHT cMYXGm"
        >
          <div
            class="sc-dlnjwi ffkZxH"
          >
            <div
              class="MuiInputBase-root MuiInput-root sc-bdnxRM eMbhKg"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                EUR
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="33bba31a"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInput-root sc-eCApnc jQvmOi Mui-disabled Mui-disabled"
            >
              <input
                class="MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled"
                disabled=""
                type="text"
                value="0"
              />
            </div>
          </div>
          <div
            class="sc-dlnjwi ffkZxH"
          >
            <span
              class="sc-hKFxyN dimflz"
            >
              Balance: 
              €
              87.00
            </span>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-jrsJWt bsREHq MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Exchange
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
